### PR TITLE
[VIRTS-3372] Optimizations to trim_links and add_test_variants

### PR DESCRIPTION
**Pull request by djlawren**
## Description

These changes intend to optimize the link generation process. Through profiling of the code it was found that the copying of link objects was the most time consuming step. Previously the `trim_links` method would create all possible variations of a link and then they would be trimmed down to only those that are valid. This created a larger amount of additional overhead later in an operation as more facts were discovered. The primary aim of these changes is to move the checks for trimming off links that are missing variables and links that do not meet requirements to before the objects are copied, thus saving a potentially significant amount of time. 

The code also reconfigures the `add_test_variants` method to hopefully make it more readable. Another small addition replaces a use of `deepcopy` with `pickle`, which has proven to be faster from an earlier change.

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Testing was performed against the existing unit tests. The changes only refactor existing functionality and not add or remove features. Code profile was compared for these changes against master when running the Thief operation with the basic fact set using the atomic planner. The mock plugin was used to run operations against ~100 agents.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by ArtificialErmine** [comment]
I need to run this when I get a chance to further verify it, but based on just what I'm seeing here, this looks like stellar work. The improvements should be handy both at the raw code and processing levels.

File: app/objects/c_agent.py [hunk 1]
```diff
@@ -270,8 +270,7 @@ async def task(self, abilities, obfuscator, facts=(), deadman=False):
             for executor in executors:
                 ex_links = [Link.load(dict(command=self.encode_string(executor.test), paw=self.paw, ability=ability,
                                            executor=executor, deadman=deadman))]
-                valid_links = await bps.remove_links_with_unset_variables(
-                    await bps.add_test_variants(links=ex_links, agent=self, facts=facts))
+                valid_links = await bps.add_test_variants(links=ex_links, agent=self, facts=facts, trim_unset_variables=True)
```
Comment (elegantmoose) [inline app/objects/c_agent.py:273]: @djlawren - Remind me, why arent we enabling 'trim_missing_requirements' here?

**Review by elegantmoose — APPROVED** [review]
Did not test locally, just static code analysis.

**Comment by elegantmoose** [comment]
@djlawren - Did you want to post the profiling test (w/ results) you executed here to see if @mchan143 or @ArtificialErmine wanted to compare with their profiling tests?

File: app/objects/c_agent.py (hunk 1 above)
Comment (djlawren) [reply app/objects/c_agent.py:273]: I do not know the reason that links missing requirements are not removed. I just noticed that the only function being called against the generated links in the original code is `remove_links_with_unset_variables`, so I only included the `trim_unset_variables` flag to maintain compatibility. I can try adding the other trim flag when I get a chance to see if it breaks any existing tests.

**Comment by djlawren** [comment]
> @djlawren - Did you want to post the profiling test (w/ results) you executed here to see if @mchan143 or @ArtificialErmine wanted to compare with their profiling tests?

Yes, I will try to post some results for this version of the code later this evening once I'm back to my laptop.

File: app/objects/c_agent.py (hunk 1 above)
Comment (djlawren) [reply app/objects/c_agent.py:273]: Oh and I just remembered that in order to trim links by missing requirements you need an `operation` object. I do not think we have access to one in this method where it's called.

**Comment by djlawren** [comment]
Here are some profiling results when running the Thief operation with the basic fact set. These results are both from the fourth iteration of the atomic scheduler. 

Here is the head of the profile of the code current as-is on master:
```
         13400388 function calls (12276368 primitive calls) in 40.150 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       86    0.002    0.000   40.134    0.467 /home/dlawrence/caldera/plugins/stockpile/app/atomic.py:42(_get_links)
       86    0.011    0.000   40.132    0.467 /home/dlawrence/caldera/app/service/planning_svc.py:140(get_links)
       86    0.013    0.000   39.508    0.459 /home/dlawrence/caldera/app/service/planning_svc.py:203(generate_and_trim_links)
       86    0.018    0.000   34.976    0.407 /home/dlawrence/caldera/app/utility/base_planning_svc.py:59(trim_links)
       86    1.326    0.015   29.980    0.349 /home/dlawrence/caldera/app/utility/base_planning_svc.py:78(add_test_variants)
    27778    5.798    0.000    7.211    0.000 {built-in method _pickle.loads}
      344    0.011    0.000    7.018    0.020 /home/dlawrence/caldera/app/utility/base_planning_svc.py:99(<listcomp>)
      430    0.061    0.000    7.007    0.016 /home/dlawrence/caldera/app/utility/base_planning_svc.py:253(_trim_by_limit)
774430/172    3.200    0.000    6.845    0.040 /home/dlawrence/.pyenv/versions/3.8.0/lib/python3.8/copy.py:128(deepcopy)
41968/172    0.223    0.000    6.842    0.040 /home/dlawrence/.pyenv/versions/3.8.0/lib/python3.8/copy.py:199(_deepcopy_list)
63382/20898    0.797    0.000    6.479    0.000 /home/dlawrence/.pyenv/versions/3.8.0/lib/python3.8/copy.py:257(_reconstruct)
41796/20898    1.390    0.000    5.837    0.000 /home/dlawrence/.pyenv/versions/3.8.0/lib/python3.8/copy.py:225(_deepcopy_dict)
    27778    4.241    0.000    5.184    0.000 {built-in method _pickle.dumps}
    27778    1.256    0.000    4.624    0.000 /home/dlawrence/caldera/app/utility/base_planning_svc.py:202(_build_single_test_variant)
       86    0.032    0.000    4.516    0.053 /home/dlawrence/caldera/app/service/planning_svc.py:335(_generate_new_links)
      430    0.003    0.000    4.369    0.010 /home/dlawrence/caldera/app/utility/base_object.py:96(load)
 6364/430    0.027    0.000    4.366    0.010 /home/dlawrence/venvs/caldera/lib/python3.8/site-packages/marshmallow/schema.py:691(load)
 6364/430    0.190    0.000    4.364    0.010 /home/dlawrence/venvs/caldera/lib/python3.8/site-packages/marshmallow/schema.py:807(_do_load)
```

Here is the head of the profile when including the changes on this branch:
```
         5186319 function calls (4942767 primitive calls) in 14.243 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       86    0.002    0.000   14.228    0.165 /home/dlawrence/caldera/plugins/stockpile/app/atomic.py:42(_get_links)
       86    0.016    0.000   14.226    0.165 /home/dlawrence/caldera/app/service/planning_svc.py:140(get_links)
       86    0.013    0.000   13.599    0.158 /home/dlawrence/caldera/app/service/planning_svc.py:203(generate_and_trim_links)
       86    0.010    0.000    9.016    0.105 /home/dlawrence/caldera/app/utility/base_planning_svc.py:59(trim_links)
       86    0.078    0.001    8.496    0.099 /home/dlawrence/caldera/app/utility/base_planning_svc.py:77(add_test_variants)
       86    0.032    0.000    4.567    0.053 /home/dlawrence/caldera/app/service/planning_svc.py:335(_generate_new_links)
      430    0.003    0.000    4.422    0.010 /home/dlawrence/caldera/app/utility/base_object.py:96(load)
 6364/430    0.025    0.000    4.419    0.010 /home/dlawrence/venvs/caldera/lib/python3.8/site-packages/marshmallow/schema.py:691(load)
 6364/430    0.193    0.000    4.416    0.010 /home/dlawrence/venvs/caldera/lib/python3.8/site-packages/marshmallow/schema.py:807(_do_load)
      344    0.251    0.001    3.931    0.011 /home/dlawrence/caldera/app/utility/base_planning_svc.py:114(<listcomp>)
    28122    0.397    0.000    3.680    0.000 /home/dlawrence/caldera/app/utility/base_planning_svc.py:170(_is_missing_requirements)
    28122    0.684    0.000    3.275    0.000 /home/dlawrence/caldera/app/utility/base_planning_svc.py:260(_do_enforcements)
```

File: app/objects/c_agent.py (hunk 1 above)
Comment (elegantmoose) [reply app/objects/c_agent.py:273]: Roger. Thats fine if just carrying on current functionality.

**Review by ddavila54 — APPROVED** [review]

**Review by wbooth — CHANGES_REQUESTED** [review]
Please add tests for your changes. ideally would be broken into a couple PRs

**Pull request closed without merging**